File: app/ui/src/lib/cad/geometry/continuous.ts
import { geometries } from '@jscad/modeling'
import type { Vec3 } from '../model/types'

export type Tessellation = Readonly<{ radialSegments?: number; pathSegments?: number; meridianSegments?: number }>
export type Asphere = Readonly<{
  curvature: number
  conic: number
  coefficients: readonly Readonly<{ order: number; value: number }>[]
}>
export type ContinuousPrimitive = 'asphericCylinder' | 'ellipsoid' | 'hyperboloid' | 'paraboloid'
export type IndexedSurface = {
  points: [number, number, number][]
  triangles: [number, number, number][]
  surfaceIndices: number[]
}
export type SurfaceEvaluation = { position: Vec3; normal: Vec3; derivativeU: Vec3; derivativeV: Vec3 }

export function positive(value: unknown, name: string): number {
  if (typeof value !== 'number' || !Number.isFinite(value) || value <= 0)
    throw new Error(`${name} must be finite and positive.`)
  return value
}

export function subdivisions(value: unknown, fallback: number, minimum = 3): number {
  if (value === undefined) return fallback
  if (typeof value !== 'number' || !Number.isInteger(value) || value < minimum)
    throw new Error(`Tessellation subdivisions must be integers >= ${minimum}.`)
  return value
}

export function normalizeAsphere(input: unknown, radius: number): Asphere {
  const value = (input ?? { curvature: 0, conic: 0, coefficients: [] }) as Asphere
  if (!Number.isFinite(value.curvature) || !Number.isFinite(value.conic) || !Array.isArray(value.coefficients))
    throw new Error('Invalid asphere definition.')
  const coefficients = value.coefficients.map((term) => ({ ...term })).sort((a, b) => a.order - b.order)
  coefficients.forEach((term, index) => {
    if (
      !Number.isInteger(term.order) ||
      term.order < 4 ||
      term.order % 2 ||
      !Number.isFinite(term.value) ||
      coefficients[index - 1]?.order === term.order
    )
      throw new Error('Asphere coefficients require unique even orders >= 4 and finite values.')
  })
  if (1 - (1 + value.conic) * value.curvature ** 2 * radius ** 2 <= 0)
    throw new Error('Asphere aperture reaches an invalid or singular conic domain.')
  return { curvature: value.curvature, conic: value.conic, coefficients }
}

export function sag(face: Asphere, radius: number): { value: number; derivative: number } {
  const q = Math.sqrt(1 - (1 + face.conic) * face.curvature ** 2 * radius ** 2)
  let value = (face.curvature * radius ** 2) / (1 + q)
  let derivative = (face.curvature * radius) / q
  for (const term of face.coefficients) {
    value += term.value * radius ** term.order
    derivative += term.order * term.value * radius ** (term.order - 1)
  }
  return { value, derivative }
}

// Each conic and monomial is monotone on a nonnegative radial interval.
// Enclose the radicand before sqrt: padding only the final sag would miss the
// square root's poor conditioning near the aperture limit.
function sagBounds(face: Asphere, low: number, high: number): [number, number] {
  const conic = (r: number): [number, number] => {
    if (face.curvature === 0 || r === 0) return [0, 0]
    const product = (1 + face.conic) * face.curvature ** 2 * r * r,
      domain = 1 - product
    const error = Number.EPSILON * 16 * (1 + Math.abs(product)) + Number.MIN_VALUE
    if (!Number.isFinite(domain) || domain - error <= 0)
      throw new Error('Asphere conic domain cannot be certified over the aperture.')
    const numerator = face.curvature * r * r,
      numeratorError = Number.EPSILON * 8 * Math.abs(numerator) + Number.MIN_VALUE
    const candidates = [
      (numerator - numeratorError) / (1 + Math.sqrt(domain - error)),
      (numerator - numeratorError) / (1 + Math.sqrt(domain + error)),
      (numerator + numeratorError) / (1 + Math.sqrt(domain - error)),
      (numerator + numeratorError) / (1 + Math.sqrt(domain + error)),
    ]
    const padding = Number.EPSILON * 8 * Math.max(...candidates.map(Math.abs)) + Number.MIN_VALUE
    return [Math.min(...candidates) - padding, Math.max(...candidates) + padding]
  }
  const first = conic(low),
    last = conic(high)
  let lower = Math.min(first[0], last[0]),
    upper = Math.max(first[1], last[1])
  for (const term of face.coefficients) {
    const a = term.value * low ** term.order,
      b = term.value * high ** term.order
    const error =
      Number.EPSILON * 32 * (term.order + 1) * (Math.abs(a) + Math.abs(b) + Math.abs(lower) + Math.abs(upper)) +
      Number.MIN_VALUE
    lower += Math.min(a, b) - error
    upper += Math.max(a, b) + error
  }
  if (!Number.isFinite(lower) || !Number.isFinite(upper))
    throw new Error('Asphere interval exceeds finite numerical range.')
  return [lower, upper]
}

export function normalizeContinuousPrimitive(
  kind: ContinuousPrimitive,
  props: Record<string, unknown>,
): Record<string, unknown> {
  for (const key of ['zMin', 'zMax', 'kind', 'focalLengths'])
    if (props[key] !== undefined) throw new Error(`${kind} does not support ${key}.`)
  if (kind === 'asphericCylinder') {
    const radius = positive(props.radius, 'radius'),
      centerThickness = positive(props.centerThickness, 'centerThickness')
    const top = normalizeAsphere(props.top, radius),
      bottom = normalizeAsphere(props.bottom, radius)
    const pending: [number, number, number][] = [[0, radius, 0]]
    let visits = 0
    while (pending.length) {
      const [lo, hi, depth] = pending.pop()!
      const t = sagBounds(top, lo, hi),
        b = sagBounds(bottom, lo, hi)
      const margin =
        Number.EPSILON *
        64 *
        (centerThickness + Math.max(Math.abs(t[0]), Math.abs(t[1])) + Math.max(Math.abs(b[0]), Math.abs(b[1])))
      if (centerThickness + t[0] - b[1] > margin) continue
      const mid = (lo + hi) / 2
      if ([lo, mid, hi].some((r) => centerThickness + sag(top, r).value - sag(bottom, r).value <= 0))
        throw new Error('AsphericCylinder surfaces intersect or touch.')
      if (++visits > 100000 || depth >= 48)
        throw new Error('AsphericCylinder positive thickness could not be certified over the full aperture.')
      pending.push([lo, mid, depth + 1], [mid, hi, depth + 1])
    }
    return { radius, centerThickness, top, bottom }
  }
  if (kind === 'paraboloid')
    return { focalLength: positive(props.focalLength, 'focalLength'), radius: positive(props.radius, 'radius') }
  const axialRadius = positive(props.axialRadius, 'axialRadius')
  const focalDistance = props.focalDistance as number
  if (
    !Number.isFinite(focalDistance) ||
    focalDistance < 0 ||
    (kind === 'ellipsoid' ? focalDistance >= axialRadius : focalDistance <= axialRadius)
  )
    throw new Error(`Invalid ${kind} focalDistance/axialRadius.`)
  return { focalDistance, axialRadius, ...(kind === 'hyperboloid' ? { radius: positive(props.radius, 'radius') } : {}) }
}

// u is polar angle / pi for Ellipsoid, radial fraction otherwise; v is azimuth.
export function evaluateContinuousSurface(
  kind: ContinuousPrimitive,
  parameters: Record<string, unknown>,
  surface: number,
  u: number,
  v: number,
): SurfaceEvaluation {
  const p = parameters as Record<string, number> & { top: Asphere; bottom: Asphere }
  const valid = kind === 'ellipsoid' ? [0] : kind === 'asphericCylinder' ? [0, 1, 2] : [1, 2]
  if (!valid.includes(surface) || !Number.isFinite(u) || u < 0 || u > 1 || !Number.isFinite(v))
    throw new Error('Invalid continuous surface coordinates.')
  const cos = Math.cos(v),
    sin = Math.sin(v)
  let dr = p.radius,
    dz = 0
  let radius = p.radius * u,
    z: number,
    normal: [number, number, number]
  if (kind === 'ellipsoid') {
    const b = Math.sqrt((p.axialRadius - p.focalDistance) * (p.axialRadius + p.focalDistance))
    radius = b * Math.sin(Math.PI * u)
    z = -p.axialRadius * Math.cos(Math.PI * u)
    dr = b * Math.PI * Math.cos(Math.PI * u)
    dz = p.axialRadius * Math.PI * Math.sin(Math.PI * u)
    normal = [(radius * cos) / b ** 2, (radius * sin) / b ** 2, z / p.axialRadius ** 2]
  } else if (kind === 'asphericCylinder') {
    if (surface === 1) {
      radius = p.radius
      const lo = -p.centerThickness / 2 + sag(p.bottom, radius).value
      const hi = p.centerThickness / 2 + sag(p.top, radius).value
      z = u === 1 ? hi : lo + (hi - lo) * u
      normal = [cos, sin, 0]
      dr = 0
      dz = hi - lo
    } else {
      const result = sag(surface === 0 ? p.bottom : p.top, radius),
        sign = surface === 0 ? -1 : 1
      z = (sign * p.centerThickness) / 2 + result.value
      normal = [-sign * result.derivative * cos, -sign * result.derivative * sin, sign]
      dz = result.derivative * p.radius
    }
  } else {
    const b2 = (p.focalDistance - p.axialRadius) * (p.focalDistance + p.axialRadius)
    const height = (r: number) =>
      kind === 'paraboloid' ? (r * r) / (4 * p.focalLength) : p.axialRadius * Math.sqrt(1 + (r * r) / b2)
    z = height(surface === 2 ? p.radius : radius)
    normal =
      surface === 2
        ? [0, 0, 1]
        : kind === 'paraboloid'
          ? [(radius * cos) / (2 * p.focalLength), (radius * sin) / (2 * p.focalLength), -1]
          : [(radius * cos) / b2, (radius * sin) / b2, -z / p.axialRadius ** 2]
    if (surface !== 2)
      dz = p.radius * (kind === 'paraboloid' ? radius / (2 * p.focalLength) : (p.axialRadius ** 2 * radius) / (b2 * z))
  }
  const length = Math.hypot(...normal)
  return {
    position: [radius * cos, radius * sin, z],
    normal: normal.map((x) => x / length) as [number, number, number],
    derivativeU: [dr * cos, dr * sin, dz],
    derivativeV: [-radius * sin, radius * cos, 0],
  }
}

export function scaleContinuousParameters(kind: ContinuousPrimitive, parameters: Record<string, unknown>, q: number) {
  positive(q, 'Length conversion factor')
  const result = { ...parameters }
  for (const key of ['radius', 'centerThickness', 'axialRadius', 'focalDistance', 'focalLength'])
    if (key in result) result[key] = (result[key] as number) * q
  if (kind === 'asphericCylinder')
    for (const key of ['top', 'bottom']) {
      const face = result[key] as Asphere
      result[key] = {
        curvature: face.curvature / q,
        conic: face.conic,
        coefficients: face.coefficients.map((term) => ({
          order: term.order,
          value: term.value * q ** (1 - term.order),
        })),
      }
    }
  return result
}

export function transformSurface(value: SurfaceEvaluation, matrix: readonly number[]): SurfaceEvaluation {
  const [a, b, c, , d, e, f, , g, h, i] = matrix
  const cofactors = [
    e * i - f * h,
    f * g - d * i,
    d * h - e * g,
    c * h - b * i,
    a * i - c * g,
    b * g - a * h,
    b * f - c * e,
    c * d - a * f,
    a * e - b * d,
  ]
  const determinant = a * cofactors[0] + b * cofactors[1] + c * cofactors[2]
  if (!Number.isFinite(determinant) || determinant === 0) throw new Error('Singular surface transform.')
  const normal = [0, 1, 2].map(
    (row) => value.normal.reduce((sum, x, j) => sum + cofactors[row * 3 + j] * x, 0) / determinant,
  )
  const length = Math.hypot(...normal)
  const vector = (v: Vec3) =>
    [0, 1, 2].map((row) => v.reduce((sum, x, j) => sum + matrix[row * 4 + j] * x, 0)) as [number, number, number]
  const position = vector(value.position).map((x, j) => x + matrix[j * 4 + 3]) as [number, number, number]
  return {
    position,
    normal: normal.map((x) => x / length) as [number, number, number],
    derivativeU: vector(value.derivativeU),
    derivativeV: vector(value.derivativeV),
  }
}

export function tessellateContinuousPrimitive(
  kind: ContinuousPrimitive,
  parameters: Record<string, unknown>,
  settings: Tessellation = {},
): IndexedSurface {
  const count = subdivisions(settings.radialSegments, 64),
    steps = subdivisions(settings.meridianSegments, 32, 2)
  const mesh: IndexedSurface = { points: [], triangles: [], surfaceIndices: [] }
  const points = new Map<string, number>()
  const vertex = (surface: number, u: number, i: number) => {
    const p = evaluateContinuousSurface(kind, parameters, surface, u, (2 * Math.PI * i) / count).position as [
      number,
      number,
      number,
    ]
    if (
      (u === 0 && surface !== 1) ||
      (kind === 'ellipsoid' && (u === 0 || u === 1)) ||
      ((kind === 'hyperboloid' || kind === 'paraboloid') && u === 0)
    )
      p[0] = p[1] = 0
    const key = p.join(',')
    let id = points.get(key)
    if (id === undefined) {
      id = mesh.points.length
      mesh.points.push(p)
      points.set(key, id)
    }
    return id
  }
  const patch = (surface: number, reversed: boolean, meridians: number) => {
    const rings = Array.from({ length: meridians + 1 }, (_, j) =>
      Array.from({ length: count }, (_, i) => vertex(surface, j / meridians, i)),
    )
    for (let j = 0; j < meridians; j++)
      for (let i = 0; i < count; i++) {
        const n = (i + 1) % count,
          a = rings[j][i],
          b = rings[j][n],
          c = rings[j + 1][n],
          d = rings[j + 1][i]
        for (const face of [
          [a, b, c],
          [a, c, d],
        ]) {
          if (new Set(face).size < 3) continue
          mesh.triangles.push((reversed ? face.reverse() : face) as [number, number, number])
          mesh.surfaceIndices.push(surface)
        }
      }
  }
  if (kind === 'ellipsoid') patch(0, false, steps)
  else if (kind === 'asphericCylinder') {
    patch(0, false, steps)
    patch(1, false, 1)
    patch(2, true, steps)
  } else {
    patch(1, false, steps)
    patch(2, true, 1)
  }
  return mesh
}

export function indexedGeometry(mesh: IndexedSurface) {
  return geometries.geom3.create(mesh.triangles.map((face) => geometries.poly3.create(face.map((i) => mesh.points[i]))))
}

export function indexedSurfaces(mesh: IndexedSurface, labels: Readonly<Record<number, string>>) {
  return Object.entries(labels)
    .map(([key, label]) => ({
      surfaceIndex: Number(key),
      label,
      polygonIndices: mesh.surfaceIndices.flatMap((index, i) => (index === Number(key) ? [i] : [])),
    }))
    .filter((surface) => surface.polygonIndices.length)
}

File: app/slaves/cae/app/methods/geometry/continuous.py
"""Continuous axial solids. Mesh resolution never changes their definitions."""

from __future__ import annotations

import math
import numpy as np


def sag(face, radius):
    c, k = face["curvature"], face["conic"]
    q = np.sqrt(1 - (1 + k) * c * c * np.asarray(radius) ** 2)
    value, derivative = (
        c * np.asarray(radius) ** 2 / (1 + q),
        c * np.asarray(radius) / q,
    )
    for term in face["coefficients"]:
        value += term["value"] * np.asarray(radius) ** term["order"]
        derivative += (
            term["order"] * term["value"] * np.asarray(radius) ** (term["order"] - 1)
        )
    return value, derivative


def evaluate_surface(kind, p, surface, u, theta):
    """u: polar fraction for ellipsoids, radial fraction otherwise (side: height)."""
    cosine, sine = np.cos(theta), np.sin(theta)
    radius = p.get("radius", 0) * u
    if kind == "ellipsoid":
        a, f = p["axialRadius"], p["focalDistance"]
        b = math.sqrt((a - f) * (a + f))
        radius, z = b * np.sin(np.pi * u), -a * np.cos(np.pi * u)
        normal = np.array([radius * cosine / b**2, radius * sine / b**2, z / a**2])
    elif kind == "asphericCylinder":
        if surface == 1:
            radius = p["radius"]
            low = -p["centerThickness"] / 2 + sag(p["bottom"], radius)[0]
            high = p["centerThickness"] / 2 + sag(p["top"], radius)[0]
            z = high if u == 1 else low + (high - low) * u
            normal = np.array([cosine, sine, 0])
        else:
            sign = -1 if surface == 0 else 1
            value, derivative = sag(p["bottom"] if surface == 0 else p["top"], radius)
            z = sign * p["centerThickness"] / 2 + value
            normal = np.array(
                [-sign * derivative * cosine, -sign * derivative * sine, sign]
            )
    elif kind in {"hyperboloid", "paraboloid"}:
        evaluated_radius = p["radius"] if surface == 2 else radius
        if kind == "paraboloid":
            z = evaluated_radius**2 / (4 * p["focalLength"])
            normal = np.array(
                [
                    radius * cosine / (2 * p["focalLength"]),
                    radius * sine / (2 * p["focalLength"]),
                    -1,
                ]
            )
        else:
            a, f = p["axialRadius"], p["focalDistance"]
            b2 = (f - a) * (f + a)
            z = a * math.sqrt(1 + evaluated_radius**2 / b2)
            normal = np.array([radius * cosine / b2, radius * sine / b2, -z / a**2])
        if surface == 2:
            normal = np.array([0.0, 0.0, 1.0])
    else:
        raise ValueError(f"Unsupported continuous primitive: {kind}")
    return np.array([radius * cosine, radius * sine, z]), normal / np.linalg.norm(
        normal
    )


def evaluate_differential(kind, p, surface, u, theta, matrix=None):
    """Continuous position, partial derivatives and inverse-transpose normal."""
    valid = (
        [0]
        if kind == "ellipsoid"
        else [0, 1, 2]
        if kind == "asphericCylinder"
        else [1, 2]
    )
    if surface not in valid or not 0 <= u <= 1 or not math.isfinite(theta):
        raise ValueError("Invalid continuous surface coordinates")
    position, normal = evaluate_surface(kind, p, surface, u, theta)
    radius = np.linalg.norm(position[:2])
    dr, dz = p.get("radius", 0), 0.0
    if kind == "ellipsoid":
        a, f = p["axialRadius"], p["focalDistance"]
        dr, dz = (
            math.sqrt((a - f) * (a + f)) * math.pi * math.cos(math.pi * u),
            a * math.pi * math.sin(math.pi * u),
        )
    elif kind == "asphericCylinder":
        if surface == 1:
            dr, dz = (
                0.0,
                p["centerThickness"]
                + sag(p["top"], radius)[0]
                - sag(p["bottom"], radius)[0],
            )
        else:
            dz = sag(p["bottom"] if surface == 0 else p["top"], radius)[1] * p["radius"]
    elif surface != 2:
        dz = p["radius"] * (
            radius / (2 * p["focalLength"])
            if kind == "paraboloid"
            else p["axialRadius"] ** 2
            * radius
            / ((p["focalDistance"] ** 2 - p["axialRadius"] ** 2) * position[2])
        )
    du = np.array([dr * math.cos(theta), dr * math.sin(theta), dz])
    dv = np.array([-radius * math.sin(theta), radius * math.cos(theta), 0.0])
    if matrix is not None:
        affine = np.asarray(matrix).reshape(4, 4)
        position = affine[:3, :3] @ position + affine[:3, 3]
        du, dv = affine[:3, :3] @ du, affine[:3, :3] @ dv
        normal = np.linalg.solve(affine[:3, :3].T, normal)
        normal /= np.linalg.norm(normal)
    return dict(position=position, normal=normal, derivativeU=du, derivativeV=dv)


def validate_primitive(kind, p):
    for key in ("zMin", "zMax", "kind", "focalLengths"):
        if key in p:
            raise ValueError(
                f"{kind} does not support {key}; migrate source and rebuild"
            )
    required = {
        "ellipsoid": ("axialRadius",),
        "hyperboloid": ("axialRadius", "radius"),
        "paraboloid": ("focalLength", "radius"),
        "asphericCylinder": ("radius", "centerThickness"),
    }
    for key in required[kind]:
        if not math.isfinite(p[key]) or p[key] <= 0:
            raise ValueError(f"{key} must be finite and positive")
    if kind in {"ellipsoid", "hyperboloid"}:
        f, a = p["focalDistance"], p["axialRadius"]
        if not math.isfinite(f) or not (0 <= f < a if kind == "ellipsoid" else f > a):
            raise ValueError("Invalid focalDistance/axialRadius")
    if kind != "asphericCylinder":
        return
    for face in (p["top"], p["bottom"]):
        c, k = face["curvature"], face["conic"]
        orders = [term["order"] for term in face["coefficients"]]
        if (
            not math.isfinite(c)
            or not math.isfinite(k)
            or 1 - (1 + k) * c * c * p["radius"] ** 2 <= 0
        ):
            raise ValueError("Invalid or singular asphere aperture")
        if (
            len(set(orders)) != len(orders)
            or any(n < 4 or n % 2 or int(n) != n for n in orders)
            or any(not math.isfinite(t["value"]) for t in face["coefficients"])
        ):
            raise ValueError("Asphere requires unique even polynomial orders >= 4")

    def bounds(face, lo, hi):
        c, k = face["curvature"], face["conic"]
        intervals = []
        for r in (lo, hi):
            if c == 0 or r == 0:
                intervals.append((0.0, 0.0))
                continue
            product = (1 + k) * c * c * r * r
            domain = 1 - product
            error = (
                np.finfo(float).eps * 16 * (1 + abs(product))
                + np.finfo(float).smallest_subnormal
            )
            if not math.isfinite(domain) or domain - error <= 0:
                raise ValueError(
                    "Asphere conic domain cannot be certified over the aperture"
                )
            numerator = c * r * r
            numerator_error = (
                np.finfo(float).eps * 8 * abs(numerator)
                + np.finfo(float).smallest_subnormal
            )
            candidates = [
                n / (1 + math.sqrt(d))
                for n in (numerator - numerator_error, numerator + numerator_error)
                for d in (domain - error, domain + error)
            ]
            padding = (
                np.finfo(float).eps * 8 * max(map(abs, candidates))
                + np.finfo(float).smallest_subnormal
            )
            intervals.append((min(candidates) - padding, max(candidates) + padding))
        lower, upper = min(x[0] for x in intervals), max(x[1] for x in intervals)
        for term in face["coefficients"]:
            a, b = (
                term["value"] * lo ** term["order"],
                term["value"] * hi ** term["order"],
            )
            error = (
                np.finfo(float).eps
                * 32
                * (term["order"] + 1)
                * (abs(a) + abs(b) + abs(lower) + abs(upper))
                + np.finfo(float).smallest_subnormal
            )
            lower += min(a, b) - error
            upper += max(a, b) + error
        if not math.isfinite(lower) or not math.isfinite(upper):
            raise ValueError("Asphere interval exceeds finite numerical range")
        return lower, upper

    pending, visits = [(0.0, p["radius"], 0)], 0
    while pending:
        lo, hi, depth = pending.pop()
        t, b = bounds(p["top"], lo, hi), bounds(p["bottom"], lo, hi)
        margin = (
            np.finfo(float).eps
            * 64
            * (p["centerThickness"] + max(map(abs, t)) + max(map(abs, b)))
        )
        if p["centerThickness"] + t[0] - b[1] > margin:
            continue
        mid = (lo + hi) / 2
        if any(
            p["centerThickness"] + sag(p["top"], r)[0] - sag(p["bottom"], r)[0] <= 0
            for r in (lo, mid, hi)
        ):
            raise ValueError("AsphericCylinder surfaces intersect or touch")
        visits += 1
        if visits > 100000 or depth >= 48:
            raise ValueError(
                "AsphericCylinder positive thickness could not be certified"
            )
        pending.extend(((lo, mid, depth + 1), (mid, hi, depth + 1)))


def tessellate_primitive(kind, parameters, settings=None):
    validate_primitive(kind, parameters)
    settings = settings or {}
    count, steps = (
        int(settings.get("radialSegments", 64)),
        int(settings.get("meridianSegments", 32)),
    )
    if (
        count < 3
        or steps < 2
        or count != settings.get("radialSegments", 64)
        or steps != settings.get("meridianSegments", 32)
    ):
        raise ValueError("Invalid tessellation subdivisions")
    points, triangles, surfaces, ids = [], [], [], {}

    def vertex(surface, u, index):
        point, _ = evaluate_surface(
            kind, parameters, surface, u, 2 * np.pi * index / count
        )
        if (
            (u == 0 and surface != 1)
            or (kind == "ellipsoid" and u in (0, 1))
            or (kind in {"hyperboloid", "paraboloid"} and u == 0)
        ):
            point[:2] = 0
        key = tuple(point)
        if key not in ids:
            ids[key] = len(points)
            points.append(point)
        return ids[key]

    def patch(surface, reverse, divisions):
        rings = [
            [vertex(surface, j / divisions, i) for i in range(count)]
            for j in range(divisions + 1)
        ]
        for j in range(divisions):
            for i in range(count):
                n = (i + 1) % count
                a, b, c, d = rings[j][i], rings[j][n], rings[j + 1][n], rings[j + 1][i]
                for face in ([a, b, c], [a, c, d]):
                    if len(set(face)) < 3:
                        continue
                    triangles.append(face[::-1] if reverse else face)
                    surfaces.append(surface)

    if kind == "ellipsoid":
        patch(0, False, steps)
    elif kind == "asphericCylinder":
        patch(0, False, steps)
        patch(1, False, 1)
        patch(2, True, steps)
    else:
        patch(1, False, steps)
        patch(2, True, 1)
    return np.asarray(points), np.asarray(triangles, dtype=np.int64), surfaces
